Federation of a custom UserRegistry (CUR) results in different behavior than when stand-alone

When federating a custom UserRegistry (CUR) using federatedRegistry-1.0, either explicitly or implicitly by including another feature often results in behavior that can break an existing Liberty configuration. When trying to reproduce the old behavior, the following federated registry configuration can be used to replicate the stand-alone CUR behavior in federatedRegistry-1.0, but may result in an error.

```
    <federatedRepository>
        <primaryRealm name="FederatedRealm">
            <participatingBaseEntry name="o=curRealmName"/>

            <groupDisplayNameMapping inputProperty="displayName" outputProperty="displayName"/>
            <uniqueGroupIdMapping inputProperty="uniqueId" outputProperty="uniqueId"/>

            <userDisplayNameMapping inputProperty="displayName" outputProperty="displayName"/>
            <uniqueUserIdMapping inputProperty="uniqueId" outputProperty="uniqueId"/>
        </primaryRealm>
    </federatedRepository>
```

The error may resemble that below (note that the REQUIRED_IDENTIFIERS_MISSING NLS message has been fixed in issue #8482):

```
Caused by: com.ibm.wsspi.security.wim.exception.WIMApplicationException: com.ibm.wsspi.security.wim.exception.WIMApplicationException: REQUIRED_IDENTIFIERS_MISSING
	at com.ibm.ws.security.wim.adapter.urbridge.utils.URBridgeEntity.setIdentifierProperties(URBridgeEntity.java:193)
	at com.ibm.ws.security.wim.adapter.urbridge.utils.URBridgeEntity.populateEntity(URBridgeEntity.java:102)
	at com.ibm.ws.security.wim.adapter.urbridge.URBridge.get(URBridge.java:458)
	... 134 more
Caused by: com.ibm.wsspi.security.wim.exception.WIMApplicationException: REQUIRED_IDENTIFIERS_MISSING
	at com.ibm.ws.security.wim.adapter.urbridge.utils.URBridgeEntity.getUniqueId(URBridgeEntity.java:216)
	at com.ibm.ws.security.wim.adapter.urbridge.utils.URBridgeEntity.setIdentifierProperties(URBridgeEntity.java:190)
	... 136 more
```

--- a/dev/com.ibm.websphere.security/src/com/ibm/websphere/security/UserRegistry.java
+++ b/dev/com.ibm.websphere.security/src/com/ibm/websphere/security/UserRegistry.java
@@ -141,7 +141,7 @@ public interface UserRegistry extends java.rmi.Remote {
                     RemoteException;
 
     /**
-     * Gets a list of users that match a <i>pattern</i> in the registy.
+     * Gets a list of users that match a <i>pattern</i> in the registry.
      * The maximum number of users returned is defined by the <i>limit</i>
      * argument.
      * <p>This method is called by GUI(adminConsole) and Scripting(Command Line) to

--- a/dev/com.ibm.ws.security.registry_test.custom/resources/OSGI-INF/l10n/metatype.properties
+++ b/dev/com.ibm.ws.security.registry_test.custom/resources/OSGI-INF/l10n/metatype.properties
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2012 IBM Corporation and others.
+# Copyright (c) 2012, 2019 IBM Corporation and others.
 # All rights reserved. This program and the accompanying materials
 # are made available under the terms of the Eclipse Public License v1.0
 # which accompanies this distribution, and is available at
@@ -17,3 +17,6 @@ usersFile.desc=The location of the file containing the registry users.
 
 groupsFile=The groups file
 groupsFile.desc=The location of the file containing the registry groups.
+
+multiValueDelimiter=The multi-value delimiter
+multiValueDelimiter.desc=The delimiter to use when multiple groups or users are defined in a single field. Useful when storing distinguished names that contain commas. Default is comma.

--- a/dev/com.ibm.ws.security.registry_test.custom/resources/OSGI-INF/metatype/metatype.xml
+++ b/dev/com.ibm.ws.security.registry_test.custom/resources/OSGI-INF/metatype/metatype.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
- Copyright (c) 2012 IBM Corporation and others.
+ Copyright (c) 2012, 2019 IBM Corporation and others.
  All rights reserved. This program and the accompanying materials
  are made available under the terms of the Eclipse Public License v1.0
  which accompanies this distribution, and is available at
@@ -17,6 +17,7 @@
          name="%fileRegistrySample.config" description="%fileRegistrySample.config.desc" >
         <AD id="usersFile" name="%usersFile" description="%usersFile.desc" required="true" type="String" />
         <AD id="groupsFile" name="%groupsFile" description="%groupsFile.desc" required="true" type="String" />
+        <AD id="multiValueDelimiter" name="%multiValueDelimiter" description="%multiValueDelimiter.desc" required="false" type="String" default="\," />
     </OCD>
 
     <Designate pid="com.ibm.ws.security.registry.custom.sample" >

--- a/dev/com.ibm.ws.security.registry_test.custom/src/com/ibm/ws/security/registry/custom/sample/FileRegistrySample.java
+++ b/dev/com.ibm.ws.security.registry_test.custom/src/com/ibm/ws/security/registry/custom/sample/FileRegistrySample.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 1997, 2016 IBM Corporation and others.
+ * Copyright (c) 1997, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -50,6 +50,7 @@ public class FileRegistrySample implements UserRegistry {
 
     private static String USERFILENAME = null;
     private static String GROUPFILENAME = null;
+    private static String MULTI_VALUE_DELIMITER = ",";
 
     /** Default Constructor **/
     public FileRegistrySample() throws java.rmi.RemoteException {}
@@ -74,21 +75,12 @@ public class FileRegistrySample implements UserRegistry {
      *                if there is any registry-specific problem
      **/
     @Override
-    public void initialize(java.util.Properties props) throws CustomRegistryException {
+    public void initialize(Properties props) throws CustomRegistryException {
         try {
-            /*
-             * try getting the USERFILENAME and the GROUPFILENAME from
-             * properties that are passed in (For example, from the
-             * administrative console). Set these values in the administrative
-             * console. Go to the special custom settings in the custom
-             * user registry section of the Authentication panel.
-             * For example:
-             * usersFile c:/temp/users.props
-             * groupsFile c:/temp/groups.props
-             */
             if (props != null) {
                 USERFILENAME = props.getProperty("usersFile");
                 GROUPFILENAME = props.getProperty("groupsFile");
+                MULTI_VALUE_DELIMITER = props.getProperty("multiValueDelimiter");
             }
 
         } catch (Exception ex) {
@@ -405,7 +397,7 @@ public class FileRegistrySample implements UserRegistry {
         }
 
         if (usrSecName == null) {
-            EntryNotFoundException ex = new EntryNotFoundException("Cannot obtain the user securityName for" + uniqueUserId);
+            EntryNotFoundException ex = new EntryNotFoundException("Cannot obtain the user securityName for " + uniqueUserId);
             throw ex;
         }
 
@@ -625,7 +617,7 @@ public class FileRegistrySample implements UserRegistry {
                     if ((s.substring(index1 + 1, index2)).equals(uniqueUserId)) {
                         int lastIndex = s.lastIndexOf(":");
                         String subs = s.substring(index2 + 1, lastIndex);
-                        StringTokenizer st1 = new StringTokenizer(subs, ",");
+                        StringTokenizer st1 = new StringTokenizer(subs, MULTI_VALUE_DELIMITER);
                         while (st1.hasMoreTokens())
                             uniqueGrpIds.add(st1.nextToken());
                         break;
@@ -747,7 +739,7 @@ public class FileRegistrySample implements UserRegistry {
                     for (int i = 0; i < 2; i++)
                         st.nextToken();
                     String subs = st.nextToken();
-                    StringTokenizer st1 = new StringTokenizer(subs, ",");
+                    StringTokenizer st1 = new StringTokenizer(subs, MULTI_VALUE_DELIMITER);
                     while (st1.hasMoreTokens()) {
                         if ((st1.nextToken()).equals(userName)) {
                             int index = s.indexOf(":");
@@ -823,7 +815,7 @@ public class FileRegistrySample implements UserRegistry {
                         for (int i = 0; i < 2; i++)
                             st.nextToken();
                         String subs = st.nextToken();
-                        StringTokenizer st1 = new StringTokenizer(subs, ",");
+                        StringTokenizer st1 = new StringTokenizer(subs, MULTI_VALUE_DELIMITER);
                         while (st1.hasMoreTokens()) {
                             user = st1.nextToken();
                             usrsForGroup.add(user);

--- a/dev/com.ibm.ws.security.registry_test.servlet/test-applications/userRegistry/src/web/UserRegistryServlet.java
+++ b/dev/com.ibm.ws.security.registry_test.servlet/test-applications/userRegistry/src/web/UserRegistryServlet.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011 IBM Corporation and others.
+ * Copyright (c) 2011, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -159,6 +159,8 @@ public class UserRegistryServlet extends HttpServlet {
         try {
             if ("getRealm".equals(method)) {
                 response = ur.getRealm();
+            } else if ("getType".equals(method)) {
+                response = ur.getType();
             } else if ("checkPassword".equals(method)) {
                 String userSecurityName = req.getParameter("userSecurityName");
                 String password = req.getParameter("password");
@@ -279,7 +281,7 @@ public class UserRegistryServlet extends HttpServlet {
      * @return The {@link List} as a String list.
      */
     private static String convertFromList(List<?> results) {
-        System.out.println("UserRegistryServlet.converFromList(): " + results.getClass() + " " + results);
+        System.out.println("UserRegistryServlet.convertFromList(): " + results.getClass() + " " + results);
 
         if (results.isEmpty()) {
             return results.toString();

--- a/dev/com.ibm.ws.security.wim.adapter.ldap_fat/build.gradle
+++ b/dev/com.ibm.ws.security.wim.adapter.ldap_fat/build.gradle
@@ -27,11 +27,14 @@ dependencies {
     project(':com.ibm.ws.com.unboundid'),
     project(':com.ibm.websphere.security.wim.base'),
     project(':com.ibm.ws.security.registry'),
+    project(':com.ibm.ws.security.registry_test.custom'),
     project(':com.ibm.ws.security.registry_test.servlet'),
     project(':com.ibm.ws.security.wim.core'),
     project(':com.ibm.ws.security.wim_test.servlet'),
     project(':com.ibm.ws.webcontainer.security_test.servlets')
 }
+
+autoFVT.dependsOn ':com.ibm.ws.security.registry_test.custom:jar'
 autoFVT.dependsOn ':com.ibm.ws.security.registry_test.servlet:war'
 autoFVT.dependsOn ':com.ibm.ws.security.wim_test.servlet:war'
 autoFVT.dependsOn ':com.ibm.ws.webcontainer.security_test.servlets:war'

--- a/dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/URAPIs_RealmPropertyMappingTest.java
+++ b/dev/com.ibm.ws.security.wim.adapter.ldap_fat/fat/src/com/ibm/ws/security/wim/adapter/ldap/fat/URAPIs_RealmPropertyMappingTest.java
@@ -42,8 +42,8 @@ import com.ibm.websphere.simplicity.config.wim.LdapRegistry;
 import com.ibm.websphere.simplicity.config.wim.RealmPropertyMapping;
 import com.ibm.websphere.simplicity.config.wim.SearchResultsCache;
 import com.ibm.websphere.simplicity.log.Log;
-import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
 import com.ibm.ws.com.unboundid.InMemoryLDAPServer;
+import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
 import com.unboundid.ldap.sdk.Entry;
 
 import componenttest.custom.junit.runner.FATRunner;

--- a/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/RepositoryManager.java
+++ b/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/RepositoryManager.java
@@ -340,13 +340,13 @@ public class RepositoryManager {
      * @param uniqueName
      * @return
      */
-    public List<String> getFederationUREntityType(String data) {
+    public List<String> getFederationUREntityType(String uniqueName) {
         List<String> result = null;
 
         for (RepositoryWrapper rw : repositories.values()) {
             if (rw instanceof UserRegistryWrapper) {
                 URBridge bridge = (URBridge) ((UserRegistryWrapper) rw).getRepository();
-                result = bridge.getEntityType(data);
+                result = bridge.getEntityType(uniqueName);
                 if (result != null) {
                     break;
                 }

--- a/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/adapter/urbridge/URBridge.java
+++ b/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/adapter/urbridge/URBridge.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2018 IBM Corporation and others.
+ * Copyright (c) 2012, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -397,6 +397,8 @@ public class URBridge implements Repository {
     public Root get(Root root) throws WIMException {
         Root returnRoot = new Root();
         String uniqueName = null;
+        String uniqueId = null;
+        String externalId = null;
         String auditName = null;
         AuditManager auditManager = new AuditManager();
 
@@ -423,12 +425,14 @@ public class URBridge implements Repository {
             List<Entity> entities = root.getEntities();
             for (Entity entity : entities) {
                 uniqueName = entity.getIdentifier().getUniqueName();
+                uniqueId = entity.getIdentifier().getUniqueId();
+                externalId = entity.getIdentifier().getExternalId();
                 if (uniqueName == null) { // find a name that we can use for audit purposes
-                    if (entity.getIdentifier().getUniqueId() != null) {
-                        auditName = entity.getIdentifier().getUniqueId();
+                    if (uniqueId != null) {
+                        auditName = uniqueId;
                     } else if (entity.getIdentifier().getExternalName() != null) {
                         auditName = entity.getIdentifier().getExternalName();
-                    } else if (entity.getIdentifier().getExternalId() != null) {
+                    } else if (externalId != null) {
                         auditName = entity.getIdentifier().getExternalId();
                     }
                 }
@@ -450,7 +454,18 @@ public class URBridge implements Repository {
                 URBridgeEntityFactory osEntityFactory = new URBridgeEntityFactory();
                 URBridgeEntity osEntity = osEntityFactory.createObject(returnEntity, this,
                                                                        propsMap, baseEntryName, entityConfigMap);
-                osEntity.setSecurityNameProp(uniqueName);
+
+                /*
+                 * To retrieve the entity, we need 1 of 2 properties set: securityName or uniqueId.
+                 */
+                if (uniqueName != null) {
+                    osEntity.setSecurityNameProp(uniqueName);
+                } else if (uniqueId != null) {
+                    osEntity.setUniqueIdProp(uniqueId);
+                } else if (externalId != null) {
+                    /* The URBridgeEntity class sets externalId from uniqueId, so do the same here. */
+                    osEntity.setUniqueIdProp(externalId);
+                }
 
                 // Get the attributes and populate the entity.
                 attrList = getAttributes(propertyCtrl, memberType);
@@ -642,6 +657,7 @@ public class URBridge implements Repository {
      * @return
      * @throws WIMException
      */
+    @FFDCIgnore({ EntryNotFoundException.class, RegistryException.class })
     private String getSecNameFromUniqueID(String uniqueId) throws WIMException {
         String METHODNAME = "getSecNameFromUniqueID";
         String secName = null;
@@ -650,7 +666,7 @@ public class URBridge implements Repository {
         } catch (EntryNotFoundException e) {
             try {
                 secName = getGroupSecurityName(uniqueId);
-            } catch (com.ibm.ws.security.registry.EntryNotFoundException renf) {
+            } catch (EntryNotFoundException renf) {
                 throw new EntityNotFoundException(WIMMessageKey.ENTITY_NOT_FOUND, Tr.formatMessage(tc, WIMMessageKey.ENTITY_NOT_FOUND,
                                                                                                    WIMMessageHelper.generateMsgParms(uniqueId)));
             } catch (RegistryException re) {

--- a/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/adapter/urbridge/utils/URBridgeEntity.java
+++ b/dev/com.ibm.ws.security.wim.core/src/com/ibm/ws/security/wim/adapter/urbridge/utils/URBridgeEntity.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2014 IBM Corporation and others.
+ * Copyright (c) 2012, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -142,6 +142,19 @@ public abstract class URBridgeEntity {
     public void setSecurityNameProp(String securityName) {
         if (securityName != null)
             entity.getIdentifier().set(securityNameProp, securityName);
+    }
+
+    /**
+     * Sets the uniqueIdProp. This is useful when creating a new object
+     * and wanting to populate its attributes. Mainly because you need 1 attribute
+     * before you can retrieve others.
+     *
+     * @param uniqueId used to set the uniqueId property
+     * @Exception an error occurred in the underlying repository.
+     */
+    public void setUniqueIdProp(String uniqueId) {
+        if (uniqueId != null)
+            entity.getIdentifier().set(uniqueIdProp, uniqueId);
     }
 
     public void setPrincipalName(String securityName) {

--- a/dev/com.ibm.ws.security.wim.registry_fat/bnd.bnd
+++ b/dev/com.ibm.ws.security.wim.registry_fat/bnd.bnd
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2017 IBM Corporation and others.
+# Copyright (c) 2017, 2019 IBM Corporation and others.
 # All rights reserved. This program and the accompanying materials
 # are made available under the terms of the Eclipse Public License v1.0
 # which accompanies this distribution, and is available at
@@ -17,6 +17,10 @@ src: \
 
 fat.project: true
 
+# Declare additional tested features not present in the original server.xml's
+tested.features: \
+    federatedRegistry-1.0
+
 -buildpath: \
 	com.ibm.ws.org.apache.directory.server;version=latest,\
 	com.ibm.websphere.javaee.servlet.3.0;version=latest,\
@@ -25,6 +29,7 @@ fat.project: true
 	com.ibm.ws.kernel.service;version=latest,\
 	com.ibm.ws.logging;version=latest,\
 	com.ibm.ws.security.registry;version=latest,\
+	com.ibm.ws.security.registry_test.custom;version=latest,\
 	com.ibm.ws.security.registry_test.servlet;version=latest,\
 	com.ibm.wsspi.org.osgi.service.component.annotations;version=latest,\
 	com.ibm.ws.com.unboundid;version=latest,\

--- a/dev/com.ibm.ws.security.wim.registry_fat/build.gradle
+++ b/dev/com.ibm.ws.security.wim.registry_fat/build.gradle
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017 IBM Corporation and others.
+ * Copyright (c) 2017, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -12,10 +12,13 @@
 dependencies { 
   requiredLibs project(':fattest.simplicity'),
     project(':com.ibm.ws.security.registry'),
+    project(':com.ibm.ws.security.registry_test.custom'),
     project(':com.ibm.ws.security.registry_test.servlet'),
     'com.unboundid:unboundid-ldapsdk:4.0.9',
     project(':com.ibm.ws.com.unboundid')
 }
+
+autoFVT.dependsOn ':com.ibm.ws.security.registry_test.custom:jar'
 autoFVT.dependsOn ':com.ibm.ws.security.registry_test.servlet:war'
 autoFVT.doLast {
   /*
@@ -34,7 +37,8 @@ autoFVT.doLast {
     'com.ibm.ws.security.wim.registry.fat.DefaultWIMRealm',
     'com.ibm.ws.security.wim.registry.fat.DefaultWIMRealmMultipleRepos',
     'com.ibm.ws.security.wim.registry.fat.WIMURRegression',
-    'com.ibm.ws.security.wim.registry.fat.MultiValuedOutputPropMapping'
+    'com.ibm.ws.security.wim.registry.fat.MultiValuedOutputPropMapping',
+    'com.ibm.ws.security.wim.registry.fat.FederateStandaloneCur'
   ]
   servers.each { server ->
     copy { 
@@ -42,5 +46,18 @@ autoFVT.doLast {
       into new File(autoFvtDir, 'publish/servers/' + server + '/dropins')
       include '*.war'
     }
+  }
+  
+    /*
+   * Copy sample custom UserRegistry feature to this test bucket.
+   */
+  copy { 
+    from project(':com.ibm.ws.security.registry_test.custom').file('publish/features/customRegistrySample-1.0.mf')
+    into new File(autoFvtDir, 'lib/LibertyFATTestFiles/internalfeatures')
+  }
+  copy {
+    from new File(project(':com.ibm.ws.security.registry_test.custom').buildDir, 'com.ibm.ws.security.registry_test.custom.jar')
+    into new File(autoFvtDir, 'lib/LibertyFATTestFiles/internalfeatures')
+    rename 'com.ibm.ws.security.registry_test.custom.jar', 'com.ibm.ws.security.registry_test.custom_1.0.jar'
   }
 }

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/DefaultWIMRealmMultipleReposTest.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/DefaultWIMRealmMultipleReposTest.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2013 IBM Corporation and others.
+ * Copyright (c) 2012, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -31,17 +31,20 @@ import com.ibm.ws.security.registry.RegistryException;
 import com.ibm.ws.security.registry.SearchResult;
 import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
 
+import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
 import componenttest.topology.impl.LibertyServer;
-import componenttest.topology.impl.LibertyServerFactory;
 import componenttest.topology.utils.LDAPUtils;
 
 @RunWith(FATRunner.class)
 @Mode(TestMode.LITE)
 public class DefaultWIMRealmMultipleReposTest {
-    private static LibertyServer server = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.registry.fat.DefaultWIMRealmMultipleRepos");
+
+    @Server("com.ibm.ws.security.wim.registry.fat.DefaultWIMRealmMultipleRepos")
+    public static LibertyServer server;
+
     private static final Class<?> c = DefaultWIMRealmMultipleReposTest.class;
     private static UserRegistryServletConnection servlet;
 

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/DefaultWIMRealmTest.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/DefaultWIMRealmTest.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2013 IBM Corporation and others.
+ * Copyright (c) 2012, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -31,17 +31,20 @@ import com.ibm.ws.security.registry.SearchResult;
 import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
 
 import componenttest.annotation.AllowedFFDC;
+import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
 import componenttest.topology.impl.LibertyServer;
-import componenttest.topology.impl.LibertyServerFactory;
 import componenttest.topology.utils.LDAPUtils;
 
 @RunWith(FATRunner.class)
 @Mode(TestMode.LITE)
 public class DefaultWIMRealmTest {
-    private static LibertyServer server = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.registry.fat.DefaultWIMRealm");
+
+    @Server("com.ibm.ws.security.wim.registry.fat.DefaultWIMRealm")
+    public static LibertyServer server;
+
     private static final Class<?> c = DefaultWIMRealmTest.class;
     private static UserRegistryServletConnection servlet;
 

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/FATSuite.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/FATSuite.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2014 IBM Corporation and others.
+ * Copyright (c) 2011, 2019 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -19,7 +19,8 @@ import org.junit.runners.Suite.SuiteClasses;
 @SuiteClasses({ DefaultWIMRealmTest.class,
                 DefaultWIMRealmMultipleReposTest.class,
                 WIMURRegressionTest.class,
-                MultiValuedOutputPropMappingTest.class
+                MultiValuedOutputPropMappingTest.class,
+                FederateStandaloneCurTest.class
 })
 /**
  * Purpose: This suite collects and runs all known good test suites.

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/FederateStandaloneCurTest.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/FederateStandaloneCurTest.java
@@ -1,0 +1,315 @@
+/*******************************************************************************
+ * Copyright (c) 2019 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.security.wim.registry.fat;
+
+import static componenttest.topology.utils.LDAPFatUtils.updateConfigDynamically;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.util.List;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.ibm.websphere.simplicity.config.ServerConfiguration;
+import com.ibm.websphere.simplicity.config.wim.FederatedRepository;
+import com.ibm.websphere.simplicity.config.wim.RealmPropertyMapping;
+import com.ibm.websphere.simplicity.log.Log;
+import com.ibm.ws.security.registry.SearchResult;
+import com.ibm.ws.security.registry.UserRegistry;
+import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
+
+import componenttest.annotation.Server;
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.custom.junit.runner.Mode;
+import componenttest.custom.junit.runner.Mode.TestMode;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.LDAPFatUtils;
+
+/**
+ * This test will test federating a stand-alone Custom UserRegistry (CUR) to federated
+ * registries, while maintaining the same behavior as the CUR had stand-alone.
+ */
+@RunWith(FATRunner.class)
+@Mode(TestMode.LITE)
+public class FederateStandaloneCurTest {
+
+    @Server("com.ibm.ws.security.wim.registry.fat.FederateStandaloneCur")
+    public static LibertyServer libertyServer;
+
+    private static final Class<?> c = FederateStandaloneCurTest.class;
+    private static UserRegistryServletConnection servlet;
+    private static ServerConfiguration startConfiguration = null;
+
+    /** The value returned from the active user registry's {@link UserRegistry#getRealm()} method when using the CUR without federatedRegistry-1.0. */
+    private static final String STANDALONE_REALM = "customRealm";
+
+    /** The value returned from the active user registry's {@link UserRegistry#getType()} method when using the CUR without federatedRegistry-1.0. */
+    private static final String STANDALONE_TYPE = "CUSTOM";
+
+    /** The value returned from the active user registry's {@link UserRegistry#getRealm()} method when using federatedRegistry-1.0. */
+    private static final String FEDERATED_REALM = "FederatedRealm";
+
+    /** The value returned from the active user registry's {@link UserRegistry#getType()} method when using federatedRegistry-1.0. */
+    private static final String FEDERATED_TYPE = "WIM";
+
+    /** Base DN for the CUR. */
+    private static final String BASE_DN = "o=" + STANDALONE_REALM;
+
+    /** The value returned from the CUR's {@link UserRegistry#getUserSecurityName(String)} method. */
+    private static final String USER_SECURITY_NAME = "uid=johndoe," + BASE_DN;
+
+    /** The value returned from the CUR's {@link UserRegistry#getUniqueUserId(String)} method. */
+    private static final String USER_UNIQUE_ID = "johndoe";
+
+    /** The value returned from the CUR's {@link UserRegistry#getUserDisplayName(String)} method. */
+    private static final String USER_DISPLAY_NAME = "John Doe";
+
+    /** The value user's password stored in the CUR. */
+    private static final String USER_PASSWORD = "password";
+
+    /** The value returned from the CUR's {@link UserRegistry#getGroupSecurityName(String)} method. */
+    private static final String GROUP_SECURITY_NAME = "cn=group1," + BASE_DN;
+
+    /** The value returned from the CUR's {@link UserRegistry#getUniqueGroupId(String)} method. */
+    private static final String GROUP_UNIQUE_ID = "group1";
+
+    /** The value returned from the CUR's {@link UserRegistry#getGroupDisplayName(String)} method. */
+    private static final String GROUP_DISPLAY_NAME = "User Group #1";
+
+    /**
+     * Setup the test case.
+     *
+     * @throws Exception If the setup failed for some reason.
+     */
+    @BeforeClass
+    public static void setupClass() throws Exception {
+        setupLibertyServer();
+        setupCustomUserRegistry();
+    }
+
+    private static void setupLibertyServer() throws Exception {
+
+        Log.info(c, "setUpLibertyServer", "Starting the server... (will wait for userRegistry servlet to start)");
+        libertyServer.copyFileToLibertyInstallRoot("lib/features", "internalfeatures/securitylibertyinternals-1.0.mf");
+        libertyServer.copyFileToLibertyInstallRoot("lib/features", "internalfeatures/customRegistrySample-1.0.mf");
+        libertyServer.copyFileToLibertyInstallRoot("lib", "internalfeatures/com.ibm.ws.security.registry_test.custom_1.0.jar");
+        libertyServer.addInstalledAppForValidation("userRegistry");
+        libertyServer.startServer(c.getName() + ".log");
+
+        /*
+         * Make sure the application has come up before proceeding
+         */
+        assertNotNull("Application userRegistry does not appear to have started.",
+                      libertyServer.waitForStringInLog("CWWKZ0001I:.*userRegistry"));
+        assertNotNull("Security service did not report it was ready",
+                      libertyServer.waitForStringInLog("CWWKS0008I"));
+        assertNotNull("Server did not came up",
+                      libertyServer.waitForStringInLog("CWWKF0011I"));
+
+        Log.info(c, "setUp", "Creating servlet connection the server");
+        servlet = new UserRegistryServletConnection(libertyServer.getHostname(), libertyServer.getHttpDefaultPort());
+
+        if (servlet.getRealm() == null) {
+            Thread.sleep(5000);
+            servlet.getRealm();
+        }
+
+        /*
+         * The original server configuration.
+         */
+        startConfiguration = libertyServer.getServerConfiguration();
+    }
+
+    /**
+     * Tear down the test.
+     */
+    @AfterClass
+    public static void teardownClass() throws Exception {
+        if (libertyServer != null) {
+            libertyServer.stopServer();
+        }
+
+        /*
+         * Delete any files we copied to the test server and uninstall the user bundles.
+         */
+        libertyServer.deleteFileFromLibertyInstallRoot("lib/features/webspheresecuritylibertyinternals-1.0.mf");
+        libertyServer.deleteFileFromLibertyInstallRoot("lib/features/customRegistrySample-1.0.mf");
+        libertyServer.deleteFileFromLibertyInstallRoot("lib/com.ibm.ws.security.registry_test.custom_1.0.jar");
+    }
+
+    /**
+     * The custom UserRegistry will contain one user and one group. It is based on some files that contains the
+     * users and groups.
+     *
+     * @throws Exception If writing to the users and groups files failed.
+     */
+    private static void setupCustomUserRegistry() throws Exception {
+
+        /**
+         * Write the custom UserRegistry users file.
+         *
+         * <pre>
+         * # Format:
+         * # name:passwd:uid:gids:display name
+         * # where name   = userId/userName of the user
+         * #       passwd = password of the user
+         * #       uid    = uniqueId of the user
+         * #       gid    = groupIds of the groups that the user belongs to
+         * #       display name = a (optional) display name for the user.
+         * </pre>
+         */
+        BufferedWriter writer = null;
+        try {
+            writer = new BufferedWriter(new FileWriter(new File(libertyServer.getServerRoot() + "/resources/security/users.props")));
+            writer.write(USER_SECURITY_NAME + ":" + USER_PASSWORD + ":" + USER_UNIQUE_ID + ":" + GROUP_UNIQUE_ID + ":" + USER_DISPLAY_NAME);
+            writer.newLine();
+        } finally {
+            writer.close();
+        }
+
+        /**
+         * Write the custom UserRegistry groups file.
+         *
+         * <pre>
+         * # Format:
+         * # name:passwd:uid:gids:display name
+         * # where name   = groupId of the group
+         * #       gid    = uniqueId of the group
+         * #       users  = list of all the userIds that the group contains
+         * #       display name = a (optional) display name for the group.
+         * </pre>
+         */
+        try {
+            writer = new BufferedWriter(new FileWriter(new File(libertyServer.getServerRoot() + "/resources/security/groups.props")));
+            writer.write(GROUP_SECURITY_NAME + ":" + GROUP_UNIQUE_ID + ":" + USER_SECURITY_NAME + ":" + GROUP_DISPLAY_NAME);
+            writer.newLine();
+        } finally {
+            writer.close();
+        }
+    }
+
+    /**
+     * Update the Liberty server configuration dynamically.
+     *
+     * @param federate Whether to enable federatedRegistry-1.0 feature.
+     * @throws Exception If the server could not be updated.
+     */
+    private static ServerConfiguration updateLibertyConfiguration(boolean federate) throws Exception {
+        ServerConfiguration server = startConfiguration.clone();
+
+        if (federate) {
+            /*
+             * Configure the federated repository.
+             */
+            server.getFeatureManager().getFeatures().add("federatedRegistry-1.0");
+            String[] participatingEntries = new String[] { BASE_DN };
+            FederatedRepository federatedRepository = LDAPFatUtils.createFederatedRepository(server, FEDERATED_REALM, participatingEntries);
+
+            // Failed with: "expected:<[johndoe]> but was:<[uid=johndoe,o=customRealm]>"
+            federatedRepository.getPrimaryRealm().setUniqueUserIdMapping(new RealmPropertyMapping("uniqueId", "uniqueId"));
+
+            // Failed with: "expected:<[John Doe]> but was:<[uid=johndoe,o=customRealm]>"
+            federatedRepository.getPrimaryRealm().setUserDisplayNameMapping(new RealmPropertyMapping("displayName", "displayName"));
+
+            // Failed with: "expected:<[User Group]> but was:<[null]>"
+            federatedRepository.getPrimaryRealm().setGroupDisplayNameMapping(new RealmPropertyMapping("displayName", "displayName"));
+
+            // Failed to find entity in getGroupSecurityName (searched for cn instead)
+            federatedRepository.getPrimaryRealm().setUniqueGroupIdMapping(new RealmPropertyMapping("uniqueId", "uniqueId"));
+        }
+
+        return server;
+    }
+
+    /**
+     * Run the test with the CUR as the active user registry (stand-alone).
+     *
+     * @throws Exception If the test fails for some unforeseen reason.
+     */
+    @Test
+    public void standaloneCustomUserRegistry() throws Exception {
+        ServerConfiguration config = updateLibertyConfiguration(false);
+        updateConfigDynamically(libertyServer, config, true);
+
+        doAssertions(STANDALONE_REALM, STANDALONE_TYPE);
+    }
+
+    /**
+     * Run the test with CUR federated in federatedRegistry-1.0.
+     *
+     * @throws Exception If the test fails for some unforeseen reason.
+     */
+    @Test
+    public void federatedCustomUserRegistry() throws Exception {
+        ServerConfiguration config = updateLibertyConfiguration(true);
+        updateConfigDynamically(libertyServer, config, true);
+
+        doAssertions(FEDERATED_REALM, FEDERATED_TYPE);
+    }
+
+    /**
+     * Perform assertions against the values returned from the active UserRegistry. The only
+     * difference between the results when calling this with the CUR as the active UserRegistry
+     * (stand-alone) and the with the CUR federated in federatedRegistry-1.0 should be the
+     * {@link UserRegistry#getType()} and {@link UserRegistry#getRealm()} values.
+     *
+     * @param realm The value expected to be returned from {@link UserRegistry#getRealm()} call.
+     * @param type The value expected to be returned from the {@link UserRegistry#getType()} call.
+     * @throws Exception If the test fails for some unforeseen reason.
+     */
+    private static void doAssertions(String realm, String type) throws Exception {
+        /*
+         * Generic CUR calls.
+         */
+        assertEquals(realm, servlet.getRealm());
+        assertEquals(type, servlet.getType());
+
+        /*
+         * UserRegistry user lookup API calls.
+         */
+        assertEquals(USER_SECURITY_NAME, servlet.checkPassword(USER_SECURITY_NAME, USER_PASSWORD));
+        assertEquals(USER_UNIQUE_ID, servlet.getUniqueUserId(USER_SECURITY_NAME));
+        assertEquals(USER_DISPLAY_NAME, servlet.getUserDisplayName(USER_SECURITY_NAME));
+        assertEquals(USER_SECURITY_NAME, servlet.getUserSecurityName(USER_UNIQUE_ID));
+
+        /*
+         * UserRegistry group lookup API calls.
+         */
+        assertEquals(GROUP_DISPLAY_NAME, servlet.getGroupDisplayName(GROUP_SECURITY_NAME));
+        assertEquals(GROUP_SECURITY_NAME, servlet.getGroupSecurityName(GROUP_UNIQUE_ID));
+        assertEquals(GROUP_UNIQUE_ID, servlet.getUniqueGroupId(GROUP_SECURITY_NAME));
+
+        /*
+         * UserRegistry search API calls.
+         */
+        SearchResult groups = servlet.getGroups("*", 0);
+        assertEquals(1, groups.getList().size());
+        assertEquals(GROUP_SECURITY_NAME, groups.getList().get(0));
+
+        List<String> groupIds = servlet.getUniqueGroupIdsForUser(USER_UNIQUE_ID);
+        assertEquals(1, groupIds.size());
+        assertEquals(GROUP_UNIQUE_ID, groupIds.get(0));
+
+        SearchResult users = servlet.getUsers("*", 0);
+        assertEquals(1, users.getList().size());
+        assertEquals(USER_SECURITY_NAME, users.getList().get(0));
+
+        SearchResult usersForGroup = servlet.getUsersForGroup(GROUP_SECURITY_NAME, 0);
+        assertEquals(1, usersForGroup.getList().size());
+        assertEquals(USER_SECURITY_NAME, usersForGroup.getList().get(0));
+    }
+}

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/MultiValuedOutputPropMappingTest.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/MultiValuedOutputPropMappingTest.java
@@ -31,11 +31,11 @@ import com.ibm.ws.com.unboundid.InMemoryLDAPServer;
 import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
 import com.unboundid.ldap.sdk.Entry;
 
+import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
 import componenttest.topology.impl.LibertyServer;
-import componenttest.topology.impl.LibertyServerFactory;
 import componenttest.topology.utils.LDAPFatUtils;
 import componenttest.topology.utils.LDAPUtils;
 
@@ -45,7 +45,10 @@ import componenttest.topology.utils.LDAPUtils;
 @RunWith(FATRunner.class)
 @Mode(TestMode.LITE)
 public class MultiValuedOutputPropMappingTest {
-    private static LibertyServer libertyServer = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.registry.fat.MultiValuedOutputPropMapping");
+
+    @Server("com.ibm.ws.security.wim.registry.fat.MultiValuedOutputPropMapping")
+    public static LibertyServer libertyServer;
+
     private static final Class<?> c = MultiValuedOutputPropMappingTest.class;
     private static UserRegistryServletConnection servlet;
     private static InMemoryLDAPServer ds = null;

--- a/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/WIMURRegressionTest.java
+++ b/dev/com.ibm.ws.security.wim.registry_fat/fat/src/com/ibm/ws/security/wim/registry/fat/WIMURRegressionTest.java
@@ -29,11 +29,11 @@ import com.ibm.ws.com.unboundid.InMemoryLDAPServer;
 import com.ibm.ws.security.registry.test.UserRegistryServletConnection;
 import com.unboundid.ldap.sdk.Entry;
 
+import componenttest.annotation.Server;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
 import componenttest.topology.impl.LibertyServer;
-import componenttest.topology.impl.LibertyServerFactory;
 import componenttest.topology.utils.LDAPFatUtils;
 import componenttest.topology.utils.LDAPUtils;
 
@@ -43,7 +43,10 @@ import componenttest.topology.utils.LDAPUtils;
 @RunWith(FATRunner.class)
 @Mode(TestMode.LITE)
 public class WIMURRegressionTest {
-    private static LibertyServer libertyServer = LibertyServerFactory.getLibertyServer("com.ibm.ws.security.wim.registry.fat.WIMURRegression");
+
+    @Server("com.ibm.ws.security.wim.registry.fat.WIMURRegression")
+    public static LibertyServer libertyServer;
+
     private static final Class<?> c = WIMURRegressionTest.class;
     private static UserRegistryServletConnection servlet;
     private static InMemoryLDAPServer ds = null;

--- a/dev/com.ibm.ws.security.wim.registry_fat/publish/servers/com.ibm.ws.security.wim.registry.fat.FederateStandaloneCur/bootstrap.properties
+++ b/dev/com.ibm.ws.security.wim.registry_fat/publish/servers/com.ibm.ws.security.wim.registry.fat.FederateStandaloneCur/bootstrap.properties
@@ -1,0 +1,3 @@
+com.ibm.ws.logging.trace.specification=*=info=enabled:com.ibm.ws.security.*=all=enabled
+com.ibm.ws.logging.max.file.size=0
+bootstrap.include=../testports.properties

--- a/dev/com.ibm.ws.security.wim.registry_fat/publish/servers/com.ibm.ws.security.wim.registry.fat.FederateStandaloneCur/server.xml
+++ b/dev/com.ibm.ws.security.wim.registry_fat/publish/servers/com.ibm.ws.security.wim.registry.fat.FederateStandaloneCur/server.xml
@@ -1,0 +1,16 @@
+<server description="FederatedStandaloneCurTest">
+	<include location="../fatTestPorts.xml" />
+
+	<!-- Enable features -->
+	<featureManager>
+		<feature>appSecurity-2.0</feature>
+		<feature>securitylibertyinternals-1.0</feature>
+		<feature>customRegistrySample-1.0</feature>
+		<feature>servlet-3.1</feature>
+	</featureManager>
+
+	<fileRegistrySample
+		usersFile="${server.config.dir}/resources/security/users.props"
+		groupsFile="${server.config.dir}/resources/security/groups.props"
+		multiValueDelimiter=";" />
+</server>

